relay-client/src/client.rs

```diff
@@ -0,0 +1,749 @@
+        self.send_internal(msg, None).await
+    }
+
+    /// Send a message and wait until the corresponding ack is received
+    pub async fn send_blocking<T: IntoPayload>(
+        &mut self,
+        msg: T,
+        timeout: Duration,
+    ) -> Result<()> {
+        let (ack_tx, ack_rx) = oneshot::channel();
+        self.send_internal(msg, Some(ack_tx)).await?;
+        match tokio::time::timeout(timeout, ack_rx).await {
+            Ok(Ok(())) => Ok(()),
+            Ok(Err(_)) => Err(eyre::eyre!("Failed to receive ack: sender dropped")),
+            Err(_) => Err(eyre::eyre!("Timeout waiting for ack")),
+        }
+    }
+
+    async fn send_internal<T: IntoPayload>(
+        &mut self,
+        msg: T,
+        ack_tx: Option<oneshot::Sender<()>>,
+    ) -> Result<()> {
+        let msg = match ack_tx {
+            Some(ack_tx) => OutgoingMessage::Blocking(msg.into_payload(), ack_tx),
+            None => OutgoingMessage::Normal(msg.into_payload()),
+        };
+        self.outgoing_tx
+            .as_ref()
+            .ok_or_eyre("client not connected")?
+            .send(msg)
+            .await
+            .inspect_err(|e| tracing::error!("Failed to send payload: {e}"))
+            .wrap_err("Failed to send payload")
+    }
+
+    /// Check if there are any pending messages
+    pub async fn has_pending_messages(&self) -> Result<usize> {
+        let command_tx = self
+            .command_tx
+            .as_ref()
+            .ok_or_else(|| eyre::eyre!("Client not connected"))?;
+        let (reply_tx, reply_rx) = oneshot::channel();
+        command_tx
+            .send(Command::GetPendingMessages(reply_tx))
+            .await?;
+        let pending_count = reply_rx.await?;
+        Ok(pending_count)
+    }
+
+    /// Request to replay pending messages
+    pub async fn replay_pending_messages(&self) -> Result<()> {
+        let command_tx = self
+            .command_tx
+            .as_ref()
+            .ok_or_else(|| eyre::eyre!("Client not connected"))?;
+        command_tx.send(Command::ReplayPendingMessages).await?;
+        Ok(())
+    }
+
+    /// Reconnect the client. On restart, pending messages will be replayed.
+    pub async fn reconnect(&self) -> Result<()> {
+        let command_tx = self
+            .command_tx
+            .as_ref()
+            .ok_or_else(|| eyre::eyre!("Client not connected"))?;
+        command_tx.send(Command::Reconnect).await?;
+        Ok(())
+    }
+
+    pub async fn graceful_shutdown(
+        &mut self,
+        wait_for_pending_messages: Duration,
+        wait_for_shutdown: Duration,
+    ) {
+        // Let's wait for all acks to be received
+        if self.has_pending_messages().await.map_or(false, |n| n > 0) {
+            tracing::info!(
+                "Giving {}ms for pending messages to be acked",
+                wait_for_pending_messages.as_millis()
+            );
+            tokio::time::sleep(wait_for_pending_messages).await;
```

Comment:
we sleep for `wait_for_pending_messages` twice. Perhaps we should divide by two here.

Reply:
nice catch! Will address in #342